pkg/pool-utils/contracts/RebalancedBasePool.sol

```diff
@@ -0,0 +1,106 @@
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-asset-manager-utils/contracts/IAssetManager.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/BalancerErrors.sol";
+
+import "./BasePool.sol";
+import "./interfaces/IRebalancedBasePoolRelayer.sol";
+
+/**
+ * @dev Base Pool associated with a relayer that guarantees it can only be joined/exited from the relayer itself.
+ * This contract as a simple mixin for pools. Implementing pools must make sure to call the BasePool's constructor
+ * properly.
+ */
+abstract contract RebalancedBasePool is BasePool {
+    using Math for uint256;
+    using Address for address;
+
+    IRebalancedBasePoolRelayer internal immutable _relayer;
+
+    constructor(IRebalancedBasePoolRelayer relayer) {
+        _require(address(relayer).isContract(), Errors.RELAYER_NOT_CONTRACT);
+        _relayer = relayer;
+    }
+
+    function getRelayer() public view returns (IRebalancedBasePoolRelayer) {
+        return _relayer;
+    }
+
+    function onJoinPool(
+        bytes32 poolId,
+        address sender,
+        address recipient,
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        bytes memory userData
+    ) public virtual override returns (uint256[] memory amountsIn, uint256[] memory dueProtocolFeeAmounts) {
+        (amountsIn, dueProtocolFeeAmounts) =
+            super.onJoinPool(poolId, sender, recipient, balances, lastChangeBlock, protocolSwapFeePercentage, userData);
+        _ensureAssetManagerBalanced(poolId, amountsIn, dueProtocolFeeAmounts, true);
+        return (amountsIn, dueProtocolFeeAmounts);
+    }
+
+    function onExitPool(
+        bytes32 poolId,
+        address sender,
+        address recipient,
+        uint256[] memory balances,
+        uint256 lastChangeBlock,
+        uint256 protocolSwapFeePercentage,
+        bytes memory userData
+    ) public virtual override returns (uint256[] memory amountsOut, uint256[] memory dueProtocolFeeAmounts) {
+        (amountsOut, dueProtocolFeeAmounts) =
+            super.onExitPool(poolId, sender, recipient, balances, lastChangeBlock, protocolSwapFeePercentage, userData);
+        _ensureAssetManagerBalanced(poolId, amountsOut, dueProtocolFeeAmounts, false);
+        return (amountsOut, dueProtocolFeeAmounts);
+    }
+
+    function _ensureAssetManagerBalanced(
+        bytes32 poolId,
+        uint256[] memory amounts,
+        uint256[] memory protocolFeeAmounts,
+        bool positive
+    ) internal view {
+        uint256 totalTokens = _getTotalTokens();
+        if (totalTokens > 0) _ensureAssetManagerBalanced(poolId, _token0, amounts[0], protocolFeeAmounts[0], positive);
```

Comment:
```suggestion
        if (totalTokens > 0) { _ensureAssetManagerBalanced(poolId, _token0, amounts[0], protocolFeeAmounts[0], positive); } else { return; }
```
Minor optimisation :shrug:

Reply:
(to be applied to all `if`s)

Reply:
Agreed from a readability perspective but isn't this preferable for gas? Granted this is very minor savings.

Reply:
I'd propose doing it in another way to make it look nicer, pls see recent changes